cpp/src/arrow/compute/exec/options.h

```diff
@@ -430,6 +430,23 @@ class ARROW_EXPORT SelectKSinkNodeOptions : public SinkNodeOptions {
   SelectKOptions select_k_options;
 };
 
+/// \brief Make a node which selects a range of rows passed through it
+///
+/// All batches pushed to this node will be accumulated, then selected, by the given
+/// fields. Then sorted batches will be forwarded to the generator in sorted order and
```

Comment:
How is FetchAndSort different from TopK?

Reply:
TopK is giving the highest K values (when sorted in descending order) right?
Fetch should be able to just pick sorted or unsorted data with an offset. So when offset=0, and it is sorted in descending order, we get TopK. I assume the Fetch node can be the most generalized node in that sense. 

The main issue with this implementation is that, it is not optimized. There is another effort to optimize in such a way that it doesn't accumulate all the data in-memory when a Fetch operation is done with no sorting.

Reply:
Do we still need TopK now that there is a fetch node?  Couldn't `TopK` just be a fetch node with `offset == 0`?

Reply:
@nealrichardson this isn't much different than the hacky end-of-plan-only top-k that we have today (and rely on for tpc-h testing).  We're mostly just extending that a little so we can adapt that to Substrait.

Reply:
I was also thinking about `TopK` it is merely a special case or we could say default of `Fetch`. I am bit concerned about the way I included the `SortOptions`. If we can agree upon that change, we could replace the rest with `Fetch`.

Reply:
Let's think about the long term for just a bit.  Once we have ordered execution I think we will only need two nodes, "sort", and "fetch".  Right now, we have three, "order_by_sink", "fetch_sink", and "select_k_sink".  So one of them will need to go away at some point and all of them will no longer be sink nodes and sorting behavior will need to be removed from "fetch".

Given that there is probably significant change coming to these nodes anyways I'm not sure how much bike shedding I want to do right now.  I think the fetch options you have today are fine.  In the future we will simply drop `FetchSinkNodeOptions::sort_options` and rename `FetchSinkNodeOptions` to `FetchNodeOptions` (and it will no longer extend `SinkNodeOptions`).